src/crimson/osd/osd_operations/snaptrim_event.cc

```diff
@@ -475,63 +475,68 @@ SnapTrimObjSubEvent::remove_or_update(
-          auto [submitted, all_completed] = pg->submit_transaction(
-            std::move(clone_obc),
-            std::move(txn),
-            std::move(osd_op_p),
-            std::move(log_entries));
-          return submitted.then_interruptible(
-            [all_completed=std::move(all_completed), this] () mutable {
-            return enter_stage<interruptor>(
-              wait_repop
-            ).then_interruptible([all_completed=std::move(all_completed)] () mutable {
-              return std::move(all_completed);
-            });
-          });
-        });
+      pp().wait_for_active
+    ).then_interruptible([this] {
+      return with_blocking_event<PGActivationBlocker::BlockingEvent,
+				 interruptor>([this] (auto&& trigger) {
+	return pg->wait_for_active_blocker.wait(std::move(trigger));
       });
-    }).handle_error_interruptible(
-      remove_or_update_iertr::pass_further{},
-      crimson::ct_error::assert_all{"unexpected error in SnapTrimObjSubEvent"}
-    );
-  });
+    }).then_interruptible([this] {
+      return enter_stage<interruptor>(
+	pp().recover_missing);
+    }).then_interruptible([] {
+      //return do_recover_missing(pg, get_target_oid());
+      return seastar::now();
+    }).then_interruptible([this] {
+      return enter_stage<interruptor>(
+	pp().get_obc);
+    }).then_interruptible([this] {
+      logger().debug("{}: getting obc for {}", *this, coid);
+      // end of commonality
+      // with_head_and_clone_obc lock both clone's and head's obcs
+      return pg->obc_loader.with_head_and_clone_obc<RWState::RWWRITE>(
+	coid,
+	[this](auto head_obc, auto clone_obc) {
+	logger().debug("{}: got clone_obc={}", *this, clone_obc->get_oid());
+	return enter_stage<interruptor>(
+	  pp().process
+	).then_interruptible(
+	  [this,clone_obc=std::move(clone_obc), head_obc=std::move(head_obc)]() mutable {
+	  logger().debug("{}: processing clone_obc={}", *this, clone_obc->get_oid());
+	  return remove_or_update(
+	    clone_obc, head_obc
+	  ).safe_then_unpack_interruptible([clone_obc, this]
+					   (auto&& txn, auto&& log_entries) mutable {
+	    auto [submitted, all_completed] = pg->submit_transaction(
+	      std::move(clone_obc),
+	      std::move(txn),
+	      std::move(osd_op_p),
+	      std::move(log_entries));
+	    return submitted.then_interruptible(
+	      [all_completed=std::move(all_completed), this] () mutable {
+	      return enter_stage<interruptor>(
+		wait_repop
+	      ).then_interruptible([all_completed=std::move(all_completed)] () mutable {
+		return std::move(all_completed);
+	      });
+	    });
+	  });
+	});
+      }).handle_error_interruptible(
+	remove_or_update_ertr::pass_further{},
+	crimson::ct_error::assert_all{"unexpected error in SnapTrimObjSubEvent"}
+      );
+    });
+  }, [this](std::exception_ptr eptr) -> snap_trim_sub_ertr::future<> {
+    logger().debug("{}: interrupted {}", *this, eptr);
+    return crimson::ct_error::eagain::make();
```

Comment:
There are two cases for the interruption, `actingset_changed` and `system_shutdown_exception`. IIUC only `actingset_changed` implies `eagain` by its name, right?